<?xml version="1.0"?>
<robot name="slamdog"
    xmlns:xacro="http://www.ros.org/wiki/xacro">
    <xacro:property name="M_PI" value="3.1415926"/>
    <xacro:property name="mass" value="1"/>
    <xacro:property name="x_length" value="0.6"/>
    <xacro:property name="y_length" value="0.8"/>
    <xacro:property name="z_length" value="0.4"/>
    <xacro:property name="wheel_length" value="0.1"/>
    <xacro:property name="wheel_radius" value="0.3"/>
    <xacro:property name="l1" value="0.4"/>
    <xacro:property name="l2" value="1"/>
    <xacro:property name="link_width" value="0.05"/>
    <xacro:property name="move_y" value="0.3"/>
    <xacro:property name="move_z" value="0.1"/>



    <material name="blue">
        <color rgba="0.0 0.5 1.0 1.0"/>
    </material>


    <material name="black">
        <color rgba="0.0 0.0 0.0 1.0"/>
    </material>

    <material name="orange">
        <color rgba="1.0 0.5 0.0 1.0"/>
    </material>

    <material name="red">
        <color rgba="1.0 0.0 0.0 1.0"/>
    </material>


    <!-- link_base -->
    <link name="link_base">
        <visual>
            <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
            <geometry>
                <box size="${x_length} ${y_length}  ${z_length}"/>
            </geometry>
            <material name="blue"/>
        </visual>


        <collision>
            <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
            <geometry>
                <box size="${x_length} ${y_length}  ${z_length}"/>
            </geometry>
        </collision>

        <visual>
            <origin xyz="-${l1/2} 0.0 -${z_length/2 - link_width/2}" rpy="0.0 0.0 0.0"/>
            <geometry>
                <box size="${l1} ${link_width}  ${link_width}"/>
            </geometry>
            <material name="blue"/>
        </visual>

        <!-- stable -->

        <visual>
            <origin xyz="${x_length/2 - link_width/2} 0.0 ${-z_length/4 - wheel_radius/2}" rpy="0.0 0.0 0.0"/>
            <geometry>
                <box size=" ${link_width}  ${link_width} ${wheel_radius - z_length/2}"/>
            </geometry>
            <material name="blue"/>
        </visual>

        <collision>
            <origin xyz="${x_length/2 - link_width/2} 0.0 ${-z_length/4 - wheel_radius/2}" rpy="0.0 0.0 0.0"/>
            <geometry>
                <box size=" ${link_width}  ${link_width} ${wheel_radius - z_length/2}"/>
            </geometry>
        </collision>

        <visual>
            <origin xyz="-${x_length/2 - link_width/2} 0.0 ${-z_length/4 - wheel_radius/2}" rpy="0.0 0.0 0.0"/>
            <geometry>
                <box size=" ${link_width}  ${link_width} ${wheel_radius - z_length/2}"/>
            </geometry>
            <material name="blue"/>
        </visual>

        <collision>
            <origin xyz="-${x_length/2 - link_width/2} 0.0 ${-z_length/4 - wheel_radius/2}" rpy="0.0 0.0 0.0"/>
            <geometry>
                <box size=" ${link_width}  ${link_width} ${wheel_radius - z_length/2}"/>
            </geometry>
        </collision>


        <inertial>
            <mass value="${mass}"/>
            <inertia ixx="1" ixy="0.0" ixz="0.0" iyy="1" iyz="0.0" izz="1"/>
        </inertial>
    </link>


    <link name="link_trailer">

        <!-- body -->
        <visual>
            <origin xyz="${-l2} 0.0 0.0" rpy="0.0 0.0 0.0"/>
            <geometry>
                <box size="${x_length} ${y_length}  ${z_length}"/>
            </geometry>
            <material name="orange"/>
        </visual>


        <collision>
            <origin xyz="${-l2} 0.0 0.0" rpy="0.0 0.0 0.0"/>
            <geometry>
                <box size="${x_length} ${y_length}  ${z_length}"/>
            </geometry>
        </collision>

        <!-- shaft -->
        <visual>
            <origin xyz="-${link_width/2} 0.0 -${z_length/2 - link_width}" rpy="0.0 0.0 0.0"/>
            <geometry>
                <box size="${link_width} ${link_width}  ${link_width*2}"/>
            </geometry>
            <material name="orange"/>
        </visual>


        <!-- <collision>
            <origin xyz="0.0 0.0 -${z_length/2 - link_width}" rpy="0.0 0.0 0.0"/>
            <geometry>
                <box size="${link_width} ${link_width}  ${link_width*2}"/>
            </geometry>
        </collision> -->

        <!-- link -->
        <visual>
            <origin xyz="${-l2/2} 0.0 -${z_length/2 - link_width/2}" rpy="0.0 0.0 0.0"/>
            <geometry>
                <box size="${l2} ${link_width}  ${link_width}"/>
            </geometry>
        </visual>

        <!-- <collision>
            <origin xyz="${-distance/2} 0.0 0.0" rpy="0.0 0.0 0.0"/>
            <geometry>
                <box size="${distance-link_width} ${link_width}  ${link_width}"/>
            </geometry>
        </collision> -->

        <inertial>
            <mass value="${mass}"/>
            <inertia ixx="1" ixy="0.0" ixz="0.0" iyy="1" iyz="0.0" izz="0.1"/>
        </inertial>
    </link>

    <!-- left wheel -->
    <link name="link_left_wheel">
        <visual>
            <origin xyz="0.0 0.0 0.0" rpy="-${M_PI/2} 0.0 0.0"/>
            <geometry>
                <cylinder radius="${wheel_radius}" length="${wheel_length}"/>
            </geometry>
            <material name="black"/>
        </visual>

        <collision>
            <origin xyz="0.0 0.0 0.0" rpy="-${M_PI/2} 0.0 0.0"/>
            <geometry>
                <cylinder radius="${wheel_radius}" length="${wheel_length}"/>
            </geometry>
        </collision>

        <inertial>
            <mass value="0.5"/>
            <inertia ixx="0.01" ixy="0.0" ixz="0.0" iyy="0.005" iyz="0.0" izz="0.005"/>
        </inertial>

    </link>

    <!-- right wheel -->
    <link name="link_right_wheel">

        <visual>
            <origin xyz="0.0 0.0 0.0" rpy="-${M_PI/2} 0.0 0.0"/>
            <geometry>
                <cylinder radius="${wheel_radius}" length="${wheel_length}"/>
            </geometry>
            <material name="black"/>
        </visual>
        <collision>
            <origin xyz="0.0 0.0 0.0" rpy="-${M_PI/2} 0.0 0.0"/>
            <geometry>
                <cylinder radius="${wheel_radius}" length="${wheel_length}"/>
            </geometry>
        </collision>
        <inertial>
            <mass value="0.5"/>
            <inertia ixx="0.01" ixy="0.0" ixz="0.0" iyy="0.005" iyz="0.0" izz="0.005"/>
        </inertial>

    </link>

    <!-- left wheel -->
    <link name="link_trailer_left_wheel">
        <visual>
            <origin xyz="0.0 0.0 0.0" rpy="-${M_PI/2} 0.0 0.0"/>
            <geometry>
                <cylinder radius="${wheel_radius}" length="${wheel_length}"/>
            </geometry>
            <material name="black"/>
        </visual>

        <collision>
            <origin xyz="0.0 0.0 0.0" rpy="-${M_PI/2} 0.0 0.0"/>
            <geometry>
                <cylinder radius="${wheel_radius}" length="${wheel_length}"/>
            </geometry>
        </collision>

        <inertial>
            <mass value="0.5"/>
            <inertia ixx="0.01" ixy="0.0" ixz="0.0" iyy="0.005" iyz="0.0" izz="0.005"/>
        </inertial>

    </link>


    <!-- right wheel -->
    <link name="link_trailer_right_wheel">

        <visual>
            <origin xyz="0.0 0.0 0.0" rpy="-${M_PI/2} 0.0 0.0"/>
            <geometry>
                <cylinder radius="${wheel_radius}" length="${wheel_length}"/>
            </geometry>
            <material name="black"/>
        </visual>
        <collision>
            <origin xyz="0.0 0.0 0.0" rpy="-${M_PI/2} 0.0 0.0"/>
            <geometry>
                <cylinder radius="${wheel_radius}" length="${wheel_length}"/>
            </geometry>
        </collision>
        <inertial>
            <mass value="0.5"/>
            <inertia ixx="0.01" ixy="0.0" ixz="0.0" iyy="0.005" iyz="0.0" izz="0.005"/>
        </inertial>

    </link>

    <!-- board -->
    <link name="board_0">
        <visual>
            <origin xyz="0.0 0.0 0.0}" rpy="0.0 ${M_PI/2} 0.0"/>
            <geometry>
                <mesh filename="package://my_aruco/urdf/markers/board_0.dae"/>
            </geometry>

        </visual>
    </link>

    <link name="board_1">
        <visual>
            <origin xyz="0.0 0.0 0.0}" rpy="0.0 ${M_PI/2} 0.0"/>
            <geometry>
                <mesh filename="package://my_aruco/urdf/markers/board_1.dae"/>
            </geometry>

        </visual>
    </link>

    <!-- board -->
    <link name="board_2">
        <visual>
            <origin xyz="0.0 0.0 0.0}" rpy="0.0 ${M_PI/2} 0.0"/>
            <geometry>
                <mesh filename="package://my_aruco/urdf/markers/board_2.dae"/>
            </geometry>

        </visual>
    </link>

    <link name="board_3">
        <visual>
            <origin xyz="0.0 0.0 0.0}" rpy="0.0 ${M_PI/2} 0.0"/>
            <geometry>
                <mesh filename="package://my_aruco/urdf/markers/board_3.dae"/>
            </geometry>

        </visual>
    </link>

    <!-- camera -->

    <link name="link_camera">
        <visual>
            <origin xyz="0.0 0.0 0.0}" rpy="0.0 0.0 0.0"/>
            <geometry>
                <box size="0.10 0.20 0.10"/>
            </geometry>
            <material name="red"/>

        </visual>
    </link>

    <!-- ==================== joint ==================== -->
    <joint name="joint_left_wheel" type="continuous">
        <origin xyz="0.0 ${y_length/2 + wheel_length/2} 0.0" rpy="0.0 0.0 0.0"/>
        <parent link="link_base"/>
        <child link="link_left_wheel"/>
        <axis xyz="0.0 1.0 0.0"/>
    </joint>

    <joint name="joint_right_wheel" type="continuous">
        <origin xyz="0.0 -${y_length/2 + wheel_length/2} 0.0" rpBluey="0.0 0.0 0.0"/>
        <parent link="link_base"/>
        <child link="link_right_wheel"/>
        <axis xyz="0.0 1.0 0.0"/>
    </joint>

    <joint name="joint_trailer_left_wheel" type="continuous">
        <origin xyz="-${l2} ${y_length/2 + wheel_length/2} 0.0" rpy="0.0 0.0 0.0"/>
        <parent link="link_trailer"/>
        <child link="link_trailer_left_wheel"/>
        <axis xyz="0.0 1.0 0.0"/>
    </joint>

    <joint name="joint_trailer_right_wheel" type="continuous">
        <origin xyz="-${l2} -${y_length/2 + wheel_length/2} 0.0" rpy="0.0 0.0 0.0"/>
        <parent link="link_trailer"/>
        <child link="link_trailer_right_wheel"/>
        <axis xyz="0.0 1.0 0.0"/>
    </joint>

    <joint name="joint_trailer" type="continuous">
        <origin xyz="-${l1} 0.0 0.0" rpy="0.0 0.0 0.0"/>
        <parent link="link_base"/>
        <child link="link_trailer"/>
        <axis xyz="0.0 0.0 1.0"/>
    </joint>

    <joint name="fix_board_0" type="fixed">
        <origin xyz="${-l2+x_length/2} ${move_y} -${z_length/2 - move_z}" rpy="0.0 0.0 0.0"/>
        <parent link="link_trailer"/>
        <child link="board_0"/>
        <!-- <axis xyz="0.0 0.0 1.0"/> -->
    </joint>

    <joint name="fix_board_1" type="fixed">
        <origin xyz="${-l2+x_length/2} ${-move_y} -${z_length/2 - move_z}" rpy="0.0 0.0 0.0"/>
        <parent link="link_trailer"/>
        <child link="board_1"/>
        <!-- <axis xyz="0.0 0.0 1.0"/> -->
    </joint>

    <joint name="fix_board_2" type="fixed">
        <origin xyz="${-l2+x_length/2} ${move_y} ${z_length/2 - move_z}" rpy="0.0 0.0 0.0"/>
        <parent link="link_trailer"/>
        <child link="board_2"/>
        <!-- <axis xyz="0.0 0.0 1.0"/> -->
    </joint>

    <joint name="fix_board_3" type="fixed">
        <origin xyz="${-l2+x_length/2} ${-move_y} ${z_length/2 - move_z}" rpy="0.0 0.0 0.0"/>
        <parent link="link_trailer"/>
        <child link="board_3"/>
        <!-- <axis xyz="0.0 0.0 1.0"/> -->
    </joint>

    <joint name="fix_camera" type="fixed">
        <origin xyz="-${x_length/2} 0.0 0.0" rpy="${M_PI} ${M_PI} 0.0"/>
        <parent link="link_base"/>
        <child link="link_camera"/>
        <!-- <axis xyz="0.0 0.0 1.0"/> -->
    </joint>

    <!-- ==================== gazeobo ==================== -->
    <gazebo reference="link_base">
        <material>Gazebo/Orange</material>
    </gazebo>

    <gazebo reference="link_trailer">
        <material>Gazebo/White</material>
    </gazebo>

    <gazebo reference="link_left_wheel">
        <material>Gazebo/Black</material>
    </gazebo>

    <gazebo reference="link_right_wheel">
        <material>Gazebo/Black</material>
    </gazebo>

    <gazebo reference="link_trailer_left_wheel">
        <material>Gazebo/Black</material>
    </gazebo>

    <gazebo reference="link_trailer_right_wheel">
        <material>Gazebo/Black</material>
    </gazebo>

    <gazebo reference="link_camera">
        <material>Gazebo/Red</material>
    </gazebo>

    <transmission name="trans_left_wheel_joint">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="joint_left_wheel">
            <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
        </joint>
        <actuator name="motor_left_wheel">
            <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
            <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>

    <transmission name="trans_right_wheel_joint">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="joint_right_wheel">
            <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
        </joint>
        <actuator name="motor_right_wheel">
            <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
            <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>

    <transmission name="trans_trailer_joint">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="joint_trailer">
            <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
        </joint>
        <actuator name="motor_trailer">
            <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
            <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>

    <gazebo>
        <plugin name="gazebo_ros_control" filename="libgazebo_ros_control.so">
            <robotNamespace>/slamdog</robotNamespace>
        </plugin>
    </gazebo>

    <!-- sensor -->


    <gazebo reference="link_camera">
        <sensor name="camera" type="camera" update_rate="30">
            <always_on>true</always_on>
            <camera>
                <horizental_fov>1.5707</horizental_fov>
                <image>
                    <width>640</width>
                    <height>480</height>
                    <format>RGB8</format>
                </image>
                <clip>
                    <near>0.01</near>
                    <far>8.0</far>
                </clip>
            </camera>
            <plugin name="camera_camera_controller" filename="libgazebo_ros_camera.so">
                <alwaysOn>true</alwaysOn>
                <updateRate>0.0</updateRate>
                <cameraName>camera</cameraName>
                <imageTopicName>image_raw</imageTopicName>
                <cameraInfoTopicName>camera_info</cameraInfoTopicName>
                <frameName>camera_link</frameName>
                <hackBaseline>0.07</hackBaseline>
                <distortionK1>0.0</distortionK1>
                <distortionK2>0.0</distortionK2>
                <distortionK3>0.0</distortionK3>
                <distortionT1>0.0</distortionT1>
                <distortionT2>0.0</distortionT2>
            </plugin>
        </sensor>

    </gazebo>

</robot>